tools/gfx/d3d12/render-d3d12.cpp

```diff
@@ -635,26 +635,54 @@ class D3D12Device : public RendererBase
         readRange.End = offset + size;
 
         void* uploadData;
-        SLANG_RETURN_ON_FAIL(buffer->m_uploadResource.getResource()->Map(
+
+        D3D12Resource uploadResource;
```

Comment:
It appears to me that `uploadResource` will be destoryed after this function, while the command on the GPU are still running, this may leads to a crash.

Reply:
The correct implementation should be to allocate a temporary `uploadResource` via `TransientResourceHeap`. There is a `TransientResourceHeap` object available via a `CommandBufferImpl` object. You can call `TransientResourceHeap::allocateStagingBuffer` to get an upload buffer.

I would modify `_uploadBufferData` to take an addiitonal `TransientResourceHeapImpl* trainsientHeap` parameter, and replace this line with `transientHeap->allocateStagingBuffer(...)`.

At the call sites of `_uploadBufferData`, just pass the transient resource heap object referenced by the command buffer.